From aaa661b33be0d172250404a49c94072dc2075023 Mon Sep 17 00:00:00 2001
From: Narendhar Reddy <quic_nare@quicinc.com>
Date: Mon, 9 Oct 2023 23:03:42 +0530
Subject: [PATCH] wifi: ath12k: Avoid unbalanced irq enablement.

In a random iteration napi_complete_done is invoking
napi_schedule and returning false, which is resulting in
unbalaced enablement of irqs.

Add a check  for the return value of napi_complete_done and
enable IRQs based on the return value.

Signed-off-by: Narendhar Reddy <quic_nare@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pcic.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/pcic.c b/drivers/net/wireless/ath/ath12k/pcic.c
index 0a7689a..de69917 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -309,8 +309,8 @@ static int ath12k_pcic_ext_grp_napi_poll(struct napi_struct *napi, int budget)
 
 	work_done = ath12k_dp_service_srng(ab, irq_grp, budget);
 	if (work_done < budget) {
-		napi_complete_done(napi, work_done);
-		ath12k_pcic_ext_grp_enable(irq_grp);
+		if(likely(napi_complete_done(napi, work_done)))
+				ath12k_pcic_ext_grp_enable(irq_grp);
 	}
 
 	if (work_done > budget)
-- 
2.17.1

